BUGFIX: Throw `PartialWorkspaceRebaseFailed` if workspace was up to date and publish didnt work

mhsdesign opened the pull request:
Resolves partially #5364

### `PartialWorkspaceRebaseFailed` exception

Thrown if the partial publish/discard cannot work because the events cannot be reordered as filtered.

This can happen for cases like attempting to publish a removal first and wanting as remaining change
a node move out of the removed descendants or publishing a node variant creation before the node is created.

We cannot reliably detect these cases in advance but in case the workspace is up-to-date its most likely such
an ordering conflict.

To solve the problem the partial operation should be retried with a different filter _or_ a full publish/discard is required.

If the workspace is outdated we cannot know for sure but suspect first that the conflict arose due to changes
in the base workspace, thus we throw `WorkspaceRebaseFailed` instead.
A forced rebase then might not solve the problem if It's because the order of events cannot be changed.
But attempting a second partial publish/discard (with up-to-date workspace) this exception will be thrown and can be reacted upon.





**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Cleanup method signature of `discardWorkspace`

mhsdesign committed:
BUGFIX: #5364 throw better exceptions if workspace was up to date and publish didnt work

kitsunet commented:
Shouldn't we merge this for now as a safety measure, not much we can do right now....

mhsdesign committed:
TASK: Introduce proper `PartialWorkspaceRebaseFailed` exception with information what exactly went wrong

The Neos Ui should be able to catch this and tell the user to publish something else first or publish all.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/5364-throw-better-exceptions-if-workspace-was-up-to-date-and-publish-didnt-work

mhsdesign committed:
TASK: Introduce test for #5364 expecting a `PartialWorkspaceRebaseFailed`

mhsdesign committed:
Hey Linter! ich komme gleich mit der rute!!!

mhsdesign commented:
Okay i introduced a proper `PartialWorkspaceRebaseFailed` exception for this reorder case, added some documentation and a test for the issue with the move node and delete;)

_now_ id say yes merge! The neos ui will still need some love making this a human readable error and also to offer a DiscardAll or PublishAll button (or maybe at some point publishAll parent nodes too ^^)

kitsunet commented:
Mmmm, the test error seems legit, I'll have a look

kitsunet committed:
Change test to expect PartialWorkspaceRebaseFailed
